neuralpredictors/measures/np_functions.py

```diff
@@ -12,10 +19,136 @@ def corr(y1, y2, axis=-1, eps=1e-8, **kwargs):
         eps:     offset to the standard deviation to avoid exploding the correlation due to small division (default 1e-8)
-        **kwargs: passed to final numpy.mean operatoin over standardized y1 * y2
+        **kwargs: passed to final numpy.mean operation over standardized y1 * y2
 
     Returns: correlation array
     """
+
     y1 = (y1 - y1.mean(axis=axis, keepdims=True)) / (y1.std(axis=axis, keepdims=True, ddof=0) + eps)
     y2 = (y2 - y2.mean(axis=axis, keepdims=True)) / (y2.std(axis=axis, keepdims=True, ddof=0) + eps)
     return (y1 * y2).mean(axis=axis, **kwargs)
+
+
+def oracle_corr_corrected(repeated_outputs: ArrayLike) -> np.ndarray:
+    """
+    Compute the corrected oracle correlations per neuron.
+
+    Args:
+        repeated_outputs (array-like): numpy array with shape (images, repeats, responses), or a list containing for each
+            image an array of shape (repeats, responses).
+
+    Returns:
+        array: Corrected oracle correlations per neuron
+    """
+
+    var_noise, var_mean = [], []
+    for repeat in repeated_outputs:
+        var_noise.append(repeat.var(axis=0))
+        var_mean.append(repeat.mean(axis=0))
+    var_noise = np.mean(np.array(var_noise), axis=0)
+    var_mean = np.var(np.array(var_mean), axis=0)
+    return var_mean / np.sqrt(var_mean * (var_mean + var_noise))
+
+
+def oracle_corr(repeated_outputs: ArrayLike) -> np.ndarray:
+    """
+    Compute the oracle correlations per neuron.
+
+    Args:
+        repeated_outputs (array-like): numpy array with shape (images, repeats, responses), or a list containing for each
+            image an array of shape (repeats, responses).
+
+    Returns:
+        array: Oracle correlations per neuron
+    """
+
+    oracles = []
+    for outputs in repeated_outputs:
+        r, n = outputs.shape
+        # compute the mean over repeats, for each neuron
+        mu = outputs.mean(axis=0, keepdims=True)
+        # compute oracle predictor
+        oracle = (mu - outputs / r) * r / (r - 1)
+
+        if np.any(np.isnan(oracle)):
+            logger.warning(
+                "{}% NaNs when calculating the oracle. NaNs will be set to Zero.".format(np.isnan(oracle).mean() * 100)
+            )
+            oracle[np.isnan(oracle)] = 0
+
+        oracles.append(oracle)
+    return corr(np.vstack(repeated_outputs), np.vstack(oracles), axis=0)
+
+
+def explainable_var(repeated_outputs: ArrayLike, eps: int = 1e-9) -> np.ndarray:
+    """
+    Compute the explainable variance per neuron.
+
+    Args:
+        repeated_outputs (array): numpy array with shape (images, repeats, responses), or a list containing for each
+            image an array of shape (repeats, responses).
+
+    Returns:
+        array: Corrected oracle correlations per neuron
+    """
+
+    img_var = []
+    total_var = np.var(np.vstack(repeated_outputs), axis=0, ddof=1)
+    for out in repeated_outputs:
+        img_var.append(np.var(out, axis=0, ddof=1))
```

Comment:
Since you already perform `np.vstack(repeated_outputs)` in the line 96, you could have just performed `var(repeated_outputs, axis=0, ddof=1)` to directly obtain `img_var`. That should be significantly more efficient than an explicit for loop.

Reply:
I think the replacement is `img_var = var(repeated_outputs, axis=1, ddof=1)` since the loop removes dimension 0 from `out`, then we're performing `var` over the "new" dimension 0 which is dimension 1 in `repeated_outputs`, right?